### PR TITLE
Implement relaxed null checks under strictEquality (SIP-79)

### PR DESCRIPTION
**Pull request by mberndt123**
Experimental implementation of relaxed null checks under strictEquality [SIP-79](https://github.com/scala/improvement-proposals/pull/133), see there for details.

It adds an experimental language option to enable the new behaviour. The implementation is largely analogous to my SIP-67 implementation, #23803.

## Have you relied on LLM-based tools in this contribution?

No

## How was the solution tested?

New automated tests

### CODE REVIEW

**Comment by SethTisue** [comment]
@mberndt123 now that SIP-79 has been accepted, then as far as I can see, there is no need for a special `relaxedNullChecks` flag for this, as strict equality is already itself experimental. this can be simply be the new behavior of the existing experimental `strictEquality` flag

unless you or @soronpo thinks it's actually desirable or necessary for this to remain separately configurable...?

**Comment by SethTisue** [comment]
@noti0na1 is it plausible you'll get to reviewing this before too much longer, or do we need to find a new reviewer...?

**Comment by soronpo** [comment]
Although strict equality is indeed a separate language flag, it currently works in production code without requiring `experimental`. Because of this, it cannot be considered as an experimental feature for this SIP to be just an addendum to it.

**Comment by SethTisue** [comment]
Oh right, I forgot that strict equality predates the establishment of "experimental" as a formal status... thanks for reminding me. It exists in this weird gray area where it's opt-in via a flag.

The question then is whether it's more like an experimental feature, or more like a standard, on-by-default feature. Personally I have leaned towards seeing it as akin to an experimental feature. The fact that it's merely opt-in is a historical accident; and I'm not sure we actually need to slow-walk changes to it...? But idk, I can see it's debatable.

It could depend partly on how widely it's used. If it's widely used despite its strange status, then maybe we do need to slow-walk changes to it? I think @fwbrasil said recently on another ticket that he's using strict equality in Kyo, for example.

**Comment by noti0na1** [comment]
> @noti0na1 is it plausible you'll get to reviewing this before too much longer, or do we need to find a new reviewer...?

I'm on vacation. I can get back to this PR next week

**Comment by fwbrasil** [comment]
Thanks for the ping @SethTisue. I've posted this request to the SIP committee: https://contributors.scala-lang.org/t/handle-strict-equality-as-stable/7516

**Comment by mberndt123** [comment]
Well, @bracevac has now confirmed in @fwbrasil's thread that strict-equality is "de facto experimental". So there is a case to be made that its behavior could be changed without further ceremony – after all, that's what experimental means, it just so happens that in this case the feature is not enabled by a proper experimental CLI flag. Also, this particular SIP isn't going to break anybody's code, it's strictly about letting some code pass that previously would have been rejected. So I'll be happy to remove the new experimental flag and just modify the existing behaviour.

Do enjoy your holidays, @noti0na1!

**Comment by fwbrasil** [comment]
Note that the SIP committee seems to have [agreed](https://github.com/scala/improvement-proposals/pull/133#issuecomment-4810887130) to move forward with this as experimental: 

> Today the SIP committee agreed to move this SIP forward to the experimental phase. Congrats!

I can't see why the process failure of the introduction of `strictEquality` would transfer to this new change.

**Comment by mberndt123** [comment]
> I can't see why the process failure of the introduction of `strictEquality` would transfer to this new change.

Because it could be argued that it's not a new change, it's a change to the existing experimental `strictEquality` behaviour. I'm not insisting, I'm just wondering if we might be able to cut a little red tape here.

File: compiler/src/dotty/tools/dotc/typer/Implicits.scala [hunk 1]
```diff
@@ -86,7 +86,10 @@ object Implicits:
   def strictEquality(using Context): Boolean =
     ctx.mode.is(Mode.StrictEquality) || Feature.enabled(nme.strictEquality)
 
-  /** A common base class of contextual implicits and of-type implicits which
+  def relaxedNullChecks(using Context): Boolean =
+    Feature.enabled(Feature.relaxedNullChecks)
+
+/** A common base class of contextual implicits and of-type implicits which
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/typer/Implicits.scala:92]: Fix comment indentation

File: compiler/src/dotty/tools/dotc/config/Feature.scala [hunk 2]
```diff
@@ -69,6 +70,7 @@ object Feature:
     (scala2macros, "Allow Scala 2 macros"),
     (dependent, "Allow dependent method types"),
     (erasedDefinitions, "Allow erased definitions"),
+    (relaxedNullChecks, "relaxed null checks under strictEquality (SIP-79)"),
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/config/Feature.scala:73]: Change the description to `Allow ...`/`Enable ..`

File: library/src/scala/language.scala [hunk 3]
```diff
@@ -251,6 +251,13 @@ object language {
     @deprecated("`strictEqualityPatternMatching` is now standard, no language import is needed", since = "3.9")
     object strictEqualityPatternMatching
 
+    /** Experimental support for relaxed null checks under strictEquality (SIP-79)
+     *
+     * @see [[https://github.com/scala/improvement-proposals/pull/133]]
```
Comment (noti0na1) [inline library/src/scala/language.scala:256]: We should update the scala document and point to it.

**Review by noti0na1 — CHANGES_REQUESTED** [review]
The changes look good to me based on the SIP description. 

Although the tests need to be extended:
1. Add more tests, especially neg tests for `relaxedNullChecks`, like reference types `String`?
2. I guess the feature is not specific for explicit nulls, so we need tests outside `explcit-nulls` as well.
3. Add a full flow-typing example to demonstrate the motivation of the SIP.
4. Add a test to explicitly show comparing with `val n: Null` is not supported.

**Comment by Gedochao** [comment]
https://contributors.scala-lang.org/t/handle-strict-equality-as-stable/7516/8
As per Scala Core decision, we will treat `strictEquality` as stable from now on.
And so, this change would be added as an experimental extension and require an appropriate setting.
